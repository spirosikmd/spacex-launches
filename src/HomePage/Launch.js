import React from 'react';
import PropTypes from 'prop-types';
import { Link } from '@reach/router';
import { withStyles } from '@material-ui/core/styles';
import Typography from '@material-ui/core/Typography';
import Button from '@material-ui/core/Button';
import Card from '@material-ui/core/Card';
import CardHeader from '@material-ui/core/CardHeader';
import CardContent from '@material-ui/core/CardContent';
import CardActions from '@material-ui/core/CardActions';
import LaunchDateTime from '../LaunchDateTime';
import LaunchStatus from '../LaunchStatus';

const styles = theme => ({
  launch: {
    display: 'flex',
    alignItems: 'center',
  },
  launchInfo: {
    flexGrow: 1,
  },
  connector: {
    minWidth: theme.spacing.unit * 2,
    border: `0.5px dashed ${theme.palette.secondary.dark}`,
    marginTop: '1px',
  },
});

const Launch = ({ classes, launch }) => {
  const hadMissionIds = launch.missionIds.length > 0;
  const hasContent = launch.details || hadMissionIds;

  return (
    <div className={classes.launch}>
      <LaunchStatus
        isSuccessful={launch.isSuccessful}
        isFailed={launch.isFailed}
        isUpcoming={launch.isUpcoming}
        isInProgress={launch.isInProgress}
      />
      <div className={classes.connector} />
      <Card className={classes.launchInfo}>
        <CardHeader
          title={launch.missionName}
          subheader={
            <LaunchDateTime
              utcDate={launch.utcDate}
              isTentative={launch.isTentative}
            />
          }
        />
        {hasContent && (
          <CardContent>
            {launch.details &&
              (hadMissionIds ? (
                <Typography component="p" gutterBottom>
                  {launch.details}
                </Typography>
              ) : (
                <Typography component="p">{launch.details}</Typography>
              ))}
            {hadMissionIds && (
              <Typography color="textSecondary" component="p">
                ID: {launch.missionIds.join(', ')}
              </Typography>
            )}
          </CardContent>
        )}
        {!launch.isTentative && (
          <CardActions>
            <Button
              size="small"
              color="primary"
              component={Link}
              to={`/launches/${launch.flightNumber}`}
            >
              explore
            </Button>
          </CardActions>
        )}
      </Card>
    </div>
  );
};

export const LaunchPropType = {
  isSuccessful: PropTypes.bool.isRequired,
  isFailed: PropTypes.bool.isRequired,
  isUpcoming: PropTypes.bool.isRequired,
  isInProgress: PropTypes.bool.isRequired,
  isTentative: PropTypes.bool.isRequired,
  missionName: PropTypes.string.isRequired,
  utcDate: PropTypes.object.isRequired,
  details: PropTypes.string,
  missionIds: PropTypes.array.isRequired,
  flightNumber: PropTypes.number.isRequired,
};

Launch.propTypes = {
  classes: PropTypes.object.isRequired,
  launch: PropTypes.shape(LaunchPropType).isRequired,
};

export default withStyles(styles)(React.memo(Launch));
