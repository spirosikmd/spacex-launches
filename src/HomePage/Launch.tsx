import React from 'react';
import PropTypes from 'prop-types';
import { Link } from '@reach/router';
import {
  withStyles,
  Theme,
  WithStyles,
  createStyles,
} from '@material-ui/core/styles';
import Typography from '@material-ui/core/Typography';
import Button, { ButtonProps } from '@material-ui/core/Button';
import Card from '@material-ui/core/Card';
import CardHeader from '@material-ui/core/CardHeader';
import CardContent from '@material-ui/core/CardContent';
import CardActions from '@material-ui/core/CardActions';
import LaunchDateTime from '../LaunchDateTime';
import LaunchStatus from '../LaunchStatus';
import LaunchMissionIds from '../LaunchMissionIds';
import { LaunchData } from '../api';

const styles = (theme: Theme) =>
  createStyles({
    launch: {
      display: 'flex',
      alignItems: 'center',
    },
    launchInfo: {
      flexGrow: 1,
    },
    connector: {
      minWidth: theme.spacing(2),
      border: `0.5px dashed ${theme.palette.secondary.dark}`,
      marginTop: '1px',
    },
  });

interface LinkButtonProps extends ButtonProps {
  to: string;
}

const LinkButton = (props: LinkButtonProps) => (
  <Button {...props} component={Link as any} />
);

interface LaunchProps extends WithStyles<typeof styles> {
  launch: LaunchData;
}

const Launch = ({ classes, launch }: LaunchProps) => {
  const hadMissionIds = launch.missionIds.length > 0;
  const hasContent = launch.details || hadMissionIds;

  return (
    <div className={classes.launch}>
      <LaunchStatus
        isSuccessful={launch.isSuccessful}
        isFailed={launch.isFailed}
        isUpcoming={launch.isUpcoming}
        isInProgress={launch.isInProgress}
      />
      <div className={classes.connector} />
      <Card className={classes.launchInfo}>
        <CardHeader
          title={launch.missionName}
          subheader={
            <LaunchDateTime
              utcDate={launch.utcDate}
              isTentative={launch.isTentative}
            />
          }
        />
        {hasContent && (
          <CardContent>
            {launch.details &&
              (hadMissionIds ? (
                <Typography component="p" gutterBottom>
                  {launch.details}
                </Typography>
              ) : (
                <Typography component="p">{launch.details}</Typography>
              ))}
            {hadMissionIds && (
              <LaunchMissionIds missionIds={launch.missionIds} />
            )}
          </CardContent>
        )}
        {!launch.isTentative && (
          <CardActions>
            <LinkButton
              size="small"
              color="primary"
              to={`/launches/${launch.flightNumber}`}
            >
              explore
            </LinkButton>
          </CardActions>
        )}
      </Card>
    </div>
  );
};

export const LaunchPropType = {
  isSuccessful: PropTypes.bool.isRequired,
  isFailed: PropTypes.bool.isRequired,
  isUpcoming: PropTypes.bool.isRequired,
  isInProgress: PropTypes.bool.isRequired,
  isTentative: PropTypes.bool.isRequired,
  missionName: PropTypes.string.isRequired,
  utcDate: PropTypes.object.isRequired,
  details: PropTypes.string,
  missionIds: PropTypes.array.isRequired,
  flightNumber: PropTypes.number.isRequired,
};

Launch.propTypes = {
  classes: PropTypes.object.isRequired,
  launch: PropTypes.shape(LaunchPropType).isRequired,
};

export default withStyles(styles)(React.memo(Launch));
