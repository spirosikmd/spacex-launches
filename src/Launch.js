import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { withStyles } from '@material-ui/core/styles';
import Grid from '@material-ui/core/Grid';
import Paper from '@material-ui/core/Paper';
import Typography from '@material-ui/core/Typography';
import GroupWorkRoundedIcon from '@material-ui/icons/GroupWorkRounded';

const styles = theme => ({
  status: {
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
    borderRadius: '50%',
    padding: theme.spacing.unit / 2,
    zIndex: '1',
  },
  statusIcon: {
    fontSize: 16,
  },
  upcoming: {
    color: theme.palette.grey['500'],
  },
  success: {
    color: theme.palette.success.main,
  },
  fail: {
    color: theme.palette.error.main,
  },
  launch: {
    display: 'flex',
    alignItems: 'center',
  },
  launchInfo: {
    padding: theme.spacing.unit * 2,
    flexGrow: 1,
  },
  connector: {
    minWidth: theme.spacing.unit * 2,
    border: `0.5px dashed ${theme.palette.secondary.dark}`,
    marginTop: '1px',
  },
});

export const Launch = ({ classes, launch }) => {
  const statusClassName = classnames(classes.status, {
    [classes.success]: launch.isSuccessful,
    [classes.fail]: launch.isFailed,
    [classes.upcoming]: launch.isUpcoming,
  });

  return (
    <div className={classes.launch}>
      <Paper className={statusClassName} elevation={1}>
        <GroupWorkRoundedIcon className={classes.statusIcon} />
      </Paper>
      <div className={classes.connector} />
      <Paper className={classes.launchInfo}>
        <Grid container spacing={8}>
          <Grid item xs={12}>
            <Typography>{launch.missionName}</Typography>
          </Grid>
          <Grid item xs={12}>
            <Typography color="textSecondary">
              {launch.utcDate.toLocaleString()}
            </Typography>
          </Grid>
          <Grid item xs={12}>
            <Typography>{launch.details}</Typography>
          </Grid>
        </Grid>
      </Paper>
    </div>
  );
};

export const LaunchPropType = {
  isSuccessful: PropTypes.bool.isRequired,
  isFailed: PropTypes.bool.isRequired,
  isUpcoming: PropTypes.bool.isRequired,
  isTentative: PropTypes.bool.isRequired,
  missionName: PropTypes.string.isRequired,
  utcDate: PropTypes.object.isRequired,
  details: PropTypes.string,
};

Launch.propTypes = {
  classes: PropTypes.object.isRequired,
  launch: PropTypes.shape(LaunchPropType).isRequired,
};

export default withStyles(styles)(React.memo(Launch));
