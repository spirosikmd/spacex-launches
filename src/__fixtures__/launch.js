export function createLaunchResponse(data) {
  return {
    flight_number: 1,
    mission_name: 'Test',
    launch_date_utc: '2019-01-01T18:16:00.000Z',
    launch_success: true,
    upcoming: false,
    is_tentative: false,
    details: 'Details',
    mission_id: ['id1'],
    links: {
      mission_patch: 'missionPatch',
    },
    ...data,
  };
}

export function createLaunch(data) {
  return {
    flightNumber: 1,
    missionName: 'Test',
    utcDate: new Date('2019-01-01'),
    isSuccessful: true,
    isFailed: false,
    isUpcoming: false,
    isInProgress: false,
    isTentative: false,
    details: 'Details',
    missionIds: ['id1'],
    missionPatch: 'missionPatch',
    ...data,
  };
}
