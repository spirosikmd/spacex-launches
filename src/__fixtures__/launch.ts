import { LaunchResponseData, LaunchData } from '../api';

export function createLaunchResponse(
  data: LaunchResponseData
): LaunchResponseData {
  return {
    flight_number: 1,
    mission_name: 'Test',
    launch_date_utc: '2019-01-01T18:16:00.000Z',
    launch_success: true,
    upcoming: false,
    is_tentative: false,
    details: 'Details',
    mission_id: ['id1'],
    links: {
      mission_patch: 'missionPatch',
    },
    ...data,
  };
}

export function createLaunch(data: LaunchData): LaunchData {
  return {
    flightNumber: 1,
    missionName: 'Test',
    utcDate: new Date('2019-01-01'),
    isSuccessful: true,
    isFailed: false,
    isUpcoming: false,
    isInProgress: false,
    isTentative: false,
    details: 'Details',
    missionIds: ['id1'],
    missionPatch: 'missionPatch',
    ...data,
  };
}
