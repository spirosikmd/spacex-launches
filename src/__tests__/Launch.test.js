import { Launch } from '../Launch';
import { createLaunch } from '../__fixtures__/launch';

describe('Launch', () => {
  let props;

  beforeEach(() => {
    props = {
      classes: {
        status: 'status',
        statusIcon: 'statusIcon',
        upcoming: 'upcoming',
        success: 'success',
        fail: 'fail',
        launch: 'launch',
        launchInfo: 'launchInfo',
        connector: 'connector',
      },
      launch: createLaunch(),
    };
  });

  describe('when a launch is successful', () => {
    it('renders it successful', () => {
      expect(shallow(Launch, props)).toMatchSnapshot();
    });
  });

  describe('when a launch is failed', () => {
    it('renders it failed', () => {
      props.launch.isFailed = true;
      props.launch.isSuccessful = false;
      expect(shallow(Launch, props)).toMatchSnapshot();
    });
  });

  describe('when a launch is upcoming', () => {
    it('renders it upcoming', () => {
      props.launch.isUpcoming = true;
      props.launch.isSuccessful = false;
      expect(shallow(Launch, props)).toMatchSnapshot();
    });
  });

  describe('when launch does not have details', () => {
    it('does not render the details section', () => {
      props.launch = createLaunch({ details: null });
      expect(shallow(Launch, props)).toMatchSnapshot();
    });
  });

  describe('when launch does not have mission ids', () => {
    it('does not render the mission ids info', () => {
      props.launch = createLaunch({ missionIds: [] });
      expect(shallow(Launch, props)).toMatchSnapshot();
    });
  });
});
