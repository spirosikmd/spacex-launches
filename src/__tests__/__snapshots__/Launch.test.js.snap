// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Launch when a launch is failed renders failed 1`] = `
<div
  className="launch"
>
  <WithStyles(Paper)
    className="status fail"
    elevation={1}
  >
    <pure(GroupWorkRoundedIcon)
      className="statusIcon"
    />
  </WithStyles(Paper)>
  <div
    className="connector"
  />
  <WithStyles(Card)
    className="launchInfo"
  >
    <WithStyles(CardHeader)
      subheader="Fri, 11 Jan 2019 15:31:00 GMT"
      title="Test"
    />
    <WithStyles(CardContent)>
      <WithStyles(Typography)
        component="p"
        gutterBottom={true}
      >
        Details
      </WithStyles(Typography)>
      <WithStyles(Typography)
        color="textSecondary"
        component="p"
      >
        ID: 
        id1
      </WithStyles(Typography)>
    </WithStyles(CardContent)>
    <WithStyles(CardActions)>
      <WithStyles(Button)
        color="primary"
        component={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "render": [Function],
          }
        }
        size="small"
        to="/launches/1"
      >
        explore
      </WithStyles(Button)>
    </WithStyles(CardActions)>
  </WithStyles(Card)>
</div>
`;

exports[`Launch when a launch is successful renders successful 1`] = `
<div
  className="launch"
>
  <WithStyles(Paper)
    className="status success"
    elevation={1}
  >
    <pure(GroupWorkRoundedIcon)
      className="statusIcon"
    />
  </WithStyles(Paper)>
  <div
    className="connector"
  />
  <WithStyles(Card)
    className="launchInfo"
  >
    <WithStyles(CardHeader)
      subheader="Fri, 11 Jan 2019 15:31:00 GMT"
      title="Test"
    />
    <WithStyles(CardContent)>
      <WithStyles(Typography)
        component="p"
        gutterBottom={true}
      >
        Details
      </WithStyles(Typography)>
      <WithStyles(Typography)
        color="textSecondary"
        component="p"
      >
        ID: 
        id1
      </WithStyles(Typography)>
    </WithStyles(CardContent)>
    <WithStyles(CardActions)>
      <WithStyles(Button)
        color="primary"
        component={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "render": [Function],
          }
        }
        size="small"
        to="/launches/1"
      >
        explore
      </WithStyles(Button)>
    </WithStyles(CardActions)>
  </WithStyles(Card)>
</div>
`;

exports[`Launch when a launch is upcoming renders upcoming 1`] = `
<div
  className="launch"
>
  <WithStyles(Paper)
    className="status upcoming"
    elevation={1}
  >
    <pure(GroupWorkRoundedIcon)
      className="statusIcon"
    />
  </WithStyles(Paper)>
  <div
    className="connector"
  />
  <WithStyles(Card)
    className="launchInfo"
  >
    <WithStyles(CardHeader)
      subheader="Fri, 11 Jan 2019 15:31:00 GMT"
      title="Test"
    />
    <WithStyles(CardContent)>
      <WithStyles(Typography)
        component="p"
        gutterBottom={true}
      >
        Details
      </WithStyles(Typography)>
      <WithStyles(Typography)
        color="textSecondary"
        component="p"
      >
        ID: 
        id1
      </WithStyles(Typography)>
    </WithStyles(CardContent)>
    <WithStyles(CardActions)>
      <WithStyles(Button)
        color="primary"
        component={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "render": [Function],
          }
        }
        size="small"
        to="/launches/1"
      >
        explore
      </WithStyles(Button)>
    </WithStyles(CardActions)>
  </WithStyles(Card)>
</div>
`;

exports[`Launch when launch does not have details does not render the details section 1`] = `
<div
  className="launch"
>
  <WithStyles(Paper)
    className="status success"
    elevation={1}
  >
    <pure(GroupWorkRoundedIcon)
      className="statusIcon"
    />
  </WithStyles(Paper)>
  <div
    className="connector"
  />
  <WithStyles(Card)
    className="launchInfo"
  >
    <WithStyles(CardHeader)
      subheader="Fri, 11 Jan 2019 15:31:00 GMT"
      title="Test"
    />
    <WithStyles(CardContent)>
      <WithStyles(Typography)
        color="textSecondary"
        component="p"
      >
        ID: 
        id1
      </WithStyles(Typography)>
    </WithStyles(CardContent)>
    <WithStyles(CardActions)>
      <WithStyles(Button)
        color="primary"
        component={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "render": [Function],
          }
        }
        size="small"
        to="/launches/1"
      >
        explore
      </WithStyles(Button)>
    </WithStyles(CardActions)>
  </WithStyles(Card)>
</div>
`;

exports[`Launch when launch does not have mission ids does not render the mission ids info 1`] = `
<div
  className="launch"
>
  <WithStyles(Paper)
    className="status success"
    elevation={1}
  >
    <pure(GroupWorkRoundedIcon)
      className="statusIcon"
    />
  </WithStyles(Paper)>
  <div
    className="connector"
  />
  <WithStyles(Card)
    className="launchInfo"
  >
    <WithStyles(CardHeader)
      subheader="Fri, 11 Jan 2019 15:31:00 GMT"
      title="Test"
    />
    <WithStyles(CardContent)>
      <WithStyles(Typography)
        component="p"
      >
        Details
      </WithStyles(Typography)>
    </WithStyles(CardContent)>
    <WithStyles(CardActions)>
      <WithStyles(Button)
        color="primary"
        component={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "render": [Function],
          }
        }
        size="small"
        to="/launches/1"
      >
        explore
      </WithStyles(Button)>
    </WithStyles(CardActions)>
  </WithStyles(Card)>
</div>
`;

exports[`Launch when launch is in progress renders in progress 1`] = `
<div
  className="launch"
>
  <WithStyles(Paper)
    className="status inProgress"
    elevation={1}
  >
    <pure(MoreHorizRoundedIcon)
      className="statusIcon"
    />
  </WithStyles(Paper)>
  <div
    className="connector"
  />
  <WithStyles(Card)
    className="launchInfo"
  >
    <WithStyles(CardHeader)
      subheader="Fri, 11 Jan 2019 15:31:00 GMT"
      title="Test"
    />
    <WithStyles(CardContent)>
      <WithStyles(Typography)
        component="p"
        gutterBottom={true}
      >
        Details
      </WithStyles(Typography)>
      <WithStyles(Typography)
        color="textSecondary"
        component="p"
      >
        ID: 
        id1
      </WithStyles(Typography)>
    </WithStyles(CardContent)>
    <WithStyles(CardActions)>
      <WithStyles(Button)
        color="primary"
        component={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "render": [Function],
          }
        }
        size="small"
        to="/launches/1"
      >
        explore
      </WithStyles(Button)>
    </WithStyles(CardActions)>
  </WithStyles(Card)>
</div>
`;

exports[`Launch when launch is tentative renders date without time 1`] = `
<div
  className="launch"
>
  <WithStyles(Paper)
    className="status success"
    elevation={1}
  >
    <pure(GroupWorkRoundedIcon)
      className="statusIcon"
    />
  </WithStyles(Paper)>
  <div
    className="connector"
  />
  <WithStyles(Card)
    className="launchInfo"
  >
    <WithStyles(CardHeader)
      subheader="Fri, 11 Jan 2019"
      title="Test"
    />
    <WithStyles(CardContent)>
      <WithStyles(Typography)
        component="p"
        gutterBottom={true}
      >
        Details
      </WithStyles(Typography)>
      <WithStyles(Typography)
        color="textSecondary"
        component="p"
      >
        ID: 
        id1
      </WithStyles(Typography)>
    </WithStyles(CardContent)>
    <WithStyles(CardActions)>
      <WithStyles(Button)
        color="primary"
        component={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "render": [Function],
          }
        }
        size="small"
        to="/launches/1"
      >
        explore
      </WithStyles(Button)>
    </WithStyles(CardActions)>
  </WithStyles(Card)>
</div>
`;
